compiler/frontend/src/mir_optimize/constant_folding.rs

```diff
@@ -70,226 +71,285 @@ impl Mir {
+            };
 
-                let Some(value) = index.to_usize().and_then(|index| list.get(index)) else {
-                    return Some(Err(format!("List access will panic because index {} is out of bounds.", index)));
-                };
-                Expression::Reference(*value)
-            }
-            BuiltinFunction::ListLength => {
-                let [list_id] = arguments else {
-                    return Some(Err("wrong number of arguments".to_string()));
-                };
+            Expression::Int(list.len().into())
+        }
+        BuiltinFunction::ListRemoveAt => return None,
+        BuiltinFunction::ListReplace => return None,
+        BuiltinFunction::Parallel => return None,
+        BuiltinFunction::Print => return None,
+        BuiltinFunction::StructGet => {
+            let [struct_, key] = arguments else { unreachable!() };
 
-                // TODO: Also catch this being called on a non-list and
-                // statically panic in that case.
-                let Expression::List(list) = visible.get(*list_id) else {
-                    return None;
-                };
+            // TODO: Also catch this being called on a non-struct and
+            // statically panic in that case.
+            let Expression::Struct(fields) = visible.get(*struct_) else {
+                return None;
+            };
 
-                Expression::Int(list.len().into())
+            // TODO: Relax this requirement. Even if not all keys are
+            // constant, we may still conclude the result of the builtin:
+            // If one key `semantically_equals` the requested one and all
+            // others definitely not, then we can still resolve that.
+            if !visible.get(*key).is_constant(visible) {
+                return None;
+            }
+            if fields
+                .iter()
+                .any(|(key, _)| !visible.get(*key).is_constant(visible))
+            {
+                return None;
             }
-            BuiltinFunction::StructGet => {
-                if arguments.len() != 2 {
-                    return Some(Err("wrong number of arguments".to_string()));
-                }
-
-                let struct_id = arguments[0];
-                let key_id = arguments[1];
 
-                // TODO: Also catch this being called on a non-struct and
-                // statically panic in that case.
-                let Expression::Struct(fields) = visible.get(struct_id) else {
-                    return None;
-                };
+            let value = fields
+                .iter()
+                .rev()
+                .find(|(k, _)| k.semantically_equals(*key, visible).unwrap_or(false))
+                .map(|(_, value)| *value);
+            if let Some(value) = value {
+                Expression::Reference(value)
+            } else {
+                return Some(Err(format!(
+                    "Struct access will panic because key {:?} isn't in there.",
```

Comment:
Did you mean to use `Display` to get the `RichIr`'s text?

```suggestion
                    "Struct access will panic because key {} isn't in there.",
```